modules/server/src/main/scala/seqexec/server/altair/AltairControllerEpics.scala

```diff
@@ -95,84 +94,142 @@ object AltairControllerEpics {
-            epicsTcs.targetFilter.setShortCircuit(TargetFilterClosed)
-        ),
-        newPos.filter(_ => newPosOk && !matrixOk && !prepMatrixOk).map(prepareMatrix)
-      ).collect { case Some(x) => x }
-
-      val pause   = (
-        L.debug("Pausing Altair guiding") *>
-          actions.sequence *> epicsTcs.targetFilter.post(DefaultTimeout) *>
-          L.debug("Altair guiding paused")
-      ).whenA(actions.nonEmpty)
+    private def pauseNgsMode(
+      position:     (Length, Length),
+      currCfg:      EpicsAltairConfig,
+      inst:         Instrument
+    )(pauseReasons: PauseConditionSet): PauseReturn[F] = {
+      // There are three reasons to stop NGS:
+      // 1. This is an unguided step
+      // 2. There is an offset too big to guide while applying it
+      // 3. The current control matrix will not be valid for the end position after applying an offset.
+
+      val guidedStep    = !pauseReasons.contains(PauseCondition.GaosGuideOff)
+      val isSmallOffset = pauseReasons.offsetO.forall(canGuideWhileOffseting(_, inst))
+      val currMatrixOk  = validateCurrentControlMatrix(currCfg, position)
+      val prepMatrixOk  = validatePreparedControlMatrix(currCfg, position)
+
+      val needsToStop = !guidedStep || !isSmallOffset || !currMatrixOk
+
+      val mustPrepareMatrix = (!currCfg.aoLoop || needsToStop) && guidedStep && !prepMatrixOk
+
+      // Prepare action will mark prepare the command but not run it. That will be triggered as part of the TCS
+      // configuration
+      val configActions = mustPrepareMatrix.option(prepareMatrix(position))
+
+      val pauseAction =
+        L.debug("Pausing Altair NGS guiding") *>
+          setCorrectionsOff *>
+          L.debug("Altair guiding NGS paused")
+
+      if (currCfg.aoLoop && needsToStop)
+        PauseReturn[F](
+          wasPaused = true,
+          pauseAction.some,
+          filterTarget = false,                                               // it was guiding, it will continue guiding, but there is an offset
+          GuideCapabilities(canGuideM2 = false, canGuideM1 = false),
+          configActions
+        )
+      else
+        PauseReturn[F](
+          wasPaused = false,
+          L.debug("Skipped pausing Altair NGS guiding").some,
+          filterTarget =
+            currCfg.aoLoop && !needsToStop && pauseReasons.offsetO.isDefined, // it was guiding, it will continue guiding, but there is an offset
+          GuideCapabilities(canGuideM2 = currCfg.aoLoop, canGuideM1 = currCfg.aoLoop),
+          configActions
+        )
 
-      needsToStop.option((newCfg, pause))
     }
 
+    private val setCorrectionsOff: F[ApplyCommandResult] =
+      epicsTcs.aoCorrect.setCorrections(CorrectionsOff) *> epicsTcs.aoCorrect.post(DefaultTimeout)
+
+    private val setCorrectionsOn: F[ApplyCommandResult] =
+      epicsTcs.aoCorrect.setCorrections(CorrectionsOn) *>
+        epicsTcs.aoCorrect.setGains(1) *>
```

Comment:
Without this, Altair will reset the adaptive gains when restarting the guiding leading to longer settle times.

Reply:
Maybe it deserves a comment and/or a constant rather than `1`